### PR TITLE
More time stretch hacks

### PR DESCRIPTION
**Pull request by Paul-Licameli**
Still more performance squeezed from time stretching.

Eliminate unnecessary array rotation with an equivalent change of phase calculation; vectorize multiplication.

On my machine in release builds, first indications are another 10% saved.

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

**Review by crsib — CHANGES_REQUESTED** [review]
Implementation of this code on Windows doesn't match other platforms.

**Comment by Paul-Licameli** [comment]
Rebased, and the Windows implementation is complete.

But the performance advantage that I saw before on Intel Mac no longer happens after f1d08d108c8ea620ee97dc3385f35c4c0e86a003

But I don't know what will happen to performance if #5328 (suitably rewritten) is combined.

**Review by crsib — APPROVED** [review]

**Comment by crsib** [comment]
Oh, wait, it failed to build, I think msvc doesn't like `1f` for some reason, I think `1.0f` would fix it

**Comment by Paul-Licameli** [comment]
Sorry, let's try CI again.

As I said, this seems only neutral for performance now on my machine, but this may vary.  I hope you can measure it.

**Comment by Paul-Licameli** [comment]
Let's try again to fix the build...

File: libraries/lib-time-and-pitch/StaffPad/VectorOps.h [hunk 1]
```diff
@@ -128,7 +121,7 @@ inline void rotate(
 inline void calcPhases(const std::complex<float>* src, float* dst, int32_t n)
```
Comment (saintmatthieu) [inline libraries/lib-time-and-pitch/StaffPad/VectorOps.h:121]: I would like to see an indication here that it's not a straightforward phase calculation of the `src` vector. I'm thinking renaming `calcPhasesAlternating`, consistently with the naming elsewhere, and also adding a comment where it is used in TimeAndPitch.cpp.

Comment (saintmatthieu) [reply libraries/lib-time-and-pitch/StaffPad/VectorOps.h:121]: A little bit of pseudo maths in the comment explaining the mathematical equivalence would also be welcome.

**Comment by saintmatthieu** [comment]
This should be tested for (near) bit exactness on different platforms and for both mono and stereo, which is a few combinations and I fear we miss something.
I would like to see the addition of a unit test for that, concretely, for some mono and some stereo file, checking in the checksum for each platform, and let the test check for equality.
Sometimes, such as in this PR, it might change, and for that case the test should also compute the RMS of the diff. We'd then check that diff manually and make sure that it is in an acceptable range, and if it's not, then do listening tests.

**Comment by crsib** [comment]
From what I observed, "straightforward" unit testing won't provide good coverage for the stretcher. It is really dependent on the input and behaves like a chaotic system a lot. I think for this case the detailed description in the commit message that clarifies the math is the only way to prove that the code is equivalent.

**Comment by saintmatthieu** [comment]
Chaotic? It is deterministic processing. That results are different on different architectures is clear, and that would mean a checksum for each.

**Comment by saintmatthieu** [comment]
Bit-exactness on a sufficiently complex input signal (a musical mix) would increase my confidence that the changes intended for refactoring or optimization only don't introduce unintentional changes, not yours? That'd be the sole purpose of this test, with no intention to estimate quality in absolute terms, only as a statement that the quality remains equal.

**Comment by Paul-Licameli** [comment]
On Intel Mac, this change makes bit-exact identical results to before, on at least one minutes long example of white noise shrunk to 75% duration.

The question is whether this saves any time in a release build.  I thought it did, but when I rebased onto  recent master, I could not reproduce the savings.  But this may vary across operating systems.

Given the ratio and duration of the input, is there dependency of the time stretching should take, on the content?  Possibly in the steps that identify peaks and propagate phases.  But I think not, in all the rest of it -- reading a track, interpolation (well skip that), forward and inverse FFT, overlap-add, writing to a track.

**Pull request closed without merging**